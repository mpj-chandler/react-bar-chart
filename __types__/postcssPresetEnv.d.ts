declare module '*.scss';
declare module 'postcss-preset-env';
