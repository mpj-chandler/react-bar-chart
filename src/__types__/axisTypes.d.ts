import { SeriesData, NumericDataPoint } from './seriesTypes';
import { Padding } from './styling';
import { Placement } from '../enums/Placement';

export interface AxisRange {
    min: number;
    max: number;
}

export interface AxisProps {
    data: SeriesData[];
    title: string;
    padding: Padding;
    config?: AxisConfig;
}

export interface AxisConfig {
    zeroIntercept: boolean;
    margin: number;
    tickPlacement: Placement;
    tickLength: number;
    titlePos?: {
        xPos: number,
        yPos: number
    }
}

export interface AxisTickProps {
    data: SeriesData[];
    padding: Padding;
    config?: AxisConfig;
}

