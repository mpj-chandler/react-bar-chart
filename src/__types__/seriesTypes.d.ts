import DataType from '../enums/DataType';
import { Padding } from './styling';
import { AxisRange } from './axisTypes';
import { Placement } from '../enums/Placement';

export interface DataPoint {
    tag?: string;
    y0?: number | null;
    y1: number | null;
}

export interface NamedDataPoint extends DataPoint {
    x: string;    
}

export interface DateIndexedDataPoint extends DataPoint {
    x: Date;
}

export interface NumericDataPoint extends DataPoint {
    x: number;
}

export interface NonNullNumericDataPoint extends NumericDataPoint {
    x: number;
}

export interface SeriesData {
    seriesName: string;
    type: {
        x: DataType,
        y: DataType,
    };
    points: Array<NamedDataPoint | DateIndexedDataPoint | NumericDataPoint | NonNullNumericDataPoint>;
}

export interface SeriesConfig {
    numSeries: number,
    seriesIndex: number,
    padding: Padding,
    dataType: DataType,
    xRange: AxisRange,
    yRange: any,
    placement: Placement
}