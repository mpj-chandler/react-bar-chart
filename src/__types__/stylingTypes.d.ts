interface Padding {
    left: number;
    right: number;
    top: number;
    bottom: number;
}

export default Padding;

