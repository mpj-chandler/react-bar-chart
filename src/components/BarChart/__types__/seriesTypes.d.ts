import DataType from "../enums/DataType";
import { Padding } from "./stylingTypes";
import { AxisRange } from "./axisTypes";
import Placement from "../enums/Placement";

export interface DataPoint {
    tag?: string;
}

export interface NamedDataPoint extends DataPoint {
    x: string;
    y: number;
}

export interface DateIndexedDataPoint extends DataPoint {
    x: Date;
    y: number;
}

export interface NumericDataPoint extends DataPoint {
    x: number | null;
    y: number | null;
}

export interface NonNullNumericDataPoint extends NumericDataPoint {
    x: number;
    y: number;
}

export interface SeriesData {
    seriesName: string;
    type: {
        x: DataType,
        y: DataType,
    };
    points: Array<NamedDataPoint | DateIndexedDataPoint | NumericDataPoint | NonNullNumericDataPoint>;
}

export interface SeriesConfig {
    numSeries: number,
    seriesIndex: number,
    padding: Padding,
    dataType: DataType,
    xRange: AxisRange,
    yRange: any,
    placement: Placement
}