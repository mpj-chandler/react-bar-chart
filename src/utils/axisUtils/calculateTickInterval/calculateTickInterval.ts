import { AxisRange } from '../../../__types__/axisTypes';

function calculateTickInterval(range: AxisRange, sensitivity: number = 0.25): number {
    if (range.max <= range.min) {
        throw new Error('Error! Range maximum has to be greater than range minimum.');
    }
    const interval = range.max - range.min;
    const power = Math.log10(interval) - 1;
    const remainder = power - Math.floor(power);

    if (remainder > sensitivity) {
        return Math.pow(10, Math.ceil(Math.log10(interval)) - 1);
    }

    return Math.pow(10, Math.round(Math.log10(interval)) - 1);
}

export default calculateTickInterval;

