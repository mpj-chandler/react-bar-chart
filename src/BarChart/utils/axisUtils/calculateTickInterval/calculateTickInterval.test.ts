import { AxisRange } from '../../../__types__/axisTypes';
import calculateTickInterval from './calculateTickInterval';

describe('calculateTickInterval', () => {
    describe('when using default sensitivity (0.25)', () => {
        it('it calculates the appropriate interval when modulus(max) or modulus(min) are less than 1', () => {
            const range: AxisRange = { min: -0.5, max: 1 };

            expect(calculateTickInterval(range)).toEqual(0.1);
        });

        it('it calculates the appropriate interval when modulus(max) or modulus(min) are less than 10', () => {
            const range: AxisRange = { min: -5, max: 10 };

            expect(calculateTickInterval(range)).toEqual(1);
        });

        it('it calculates the appropriate interval when modulus(max) or modulus(min) are less than 100', () => {
            const range: AxisRange = { min: -95, max: 1 };

            expect(calculateTickInterval(range)).toEqual(10);
        });

        it('it calculates the appropriate interval when modulus(max) or modulus(min) are less than 1000', () => {
            const range: AxisRange = { min: -95, max: 862 };

            expect(calculateTickInterval(range)).toEqual(100);
        });

        it('it calculates the appropriate interval when modulus(max) or modulus(min) are less than 10000', () => {
            const range: AxisRange = { min: -9500, max: 2 };

            expect(calculateTickInterval(range)).toEqual(1000);
        });

        it('it calculates the appropriate interval when modulus(max) or modulus(min) are less than 100000', () => {
            const range: AxisRange = { min: -95, max: 65987 };

            expect(calculateTickInterval(range)).toEqual(10000);
        });

        it('it throws an error when the range maximum is equal to the range minimum', () => {
            const range: AxisRange = { min: -5, max: -5 };

            expect(() => calculateTickInterval(range)).toThrowErrorMatchingSnapshot();
        });

        it('it throws an error when the range maximum is less than the range minimum', () => {
            const range: AxisRange = { min: -5, max: -50 };

            expect(() => calculateTickInterval(range)).toThrowErrorMatchingSnapshot();
        });
    });

    describe('when using custom sensitivity', () => {

        for (let sensitivity = 0; sensitivity < 1; sensitivity += 0.1) {
            it(`it calculates the appropriate interval when modulus(max) or modulus(min) are less than 100 and sensitivity is ${sensitivity}.`, () => {
                const range: AxisRange = { min: -10, max: 10 };

                expect(calculateTickInterval(range, sensitivity)).toEqual(sensitivity >= 0.4 ? 1 : 10);
            });
        }
    });
});
