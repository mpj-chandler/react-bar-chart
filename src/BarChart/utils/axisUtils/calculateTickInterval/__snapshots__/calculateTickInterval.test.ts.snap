// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculateTickInterval when using default sensitivity (0.25) it throws an error when the range maximum is equal to the range minimum 1`] = `"Error! Range maximum has to be greater than range minimum."`;

exports[`calculateTickInterval when using default sensitivity (0.25) it throws an error when the range maximum is less than the range minimum 1`] = `"Error! Range maximum has to be greater than range minimum."`;
